recipes-kernel/linux/4.14/patches/efi-tables-for-dom0.patch

```diff
@@ -0,0 +1,156 @@
+With our 32-bit dom0, it is difficult to get access to the EFI tables from OEM's
+like HP and Intel natively.  Xen already exposes a hypercall to present the EFI
+tables to a PV linux guest (ie. dom0).  This patch force compiles the support
+normally enabled w/CONFIG_XEN_EFI (requires CONFIG_X86_64) into our 32-bit dom0.
+It then selectively ignores the 32-bit/64-bit distinction during Linux kernel
+EFI subsystem initialization. The data returned from the hypercall is 32-bit
+addressable, the only unknown is whether the EFI tables themselves are 32-bit
+addressable. The EFI initialization code protects against this case by checking 
+to see if each table is below 4GB in memory, and if not, it disables EFI.
+
+This allows the sysfs DMI and ACPI entries to be populated, which dmidecode,
+xcpmd, and get-config-key rely upon.
+
+################################################################################
+CHANGELOG 
+################################################################################
+Original author: Chris Rogers <rogersc@ainfosec.com>
+
+################################################################################
+REMOVAL 
+################################################################################
+Once dom0 is 64-bit this patch can be removed.
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+None.
+
+################################################################################
+INTERNAL DEPENDENCIES 
+################################################################################
+None.
+
+################################################################################
+PATCHES 
+################################################################################
+
+--- a/arch/x86/xen/Makefile
++++ b/arch/x86/xen/Makefile
+@@ -16,7 +16,7 @@ CFLAGS_mmu_pv.o		:= $(nostackp)
+ 
+ obj-y		:= enlighten.o multicalls.o mmu.o irq.o \
+ 			time.o xen-asm.o xen-asm_$(BITS).o \
+-			grant-table.o suspend.o platform-pci-unplug.o
++			grant-table.o suspend.o platform-pci-unplug.o efi.o
+ 
+ obj-$(CONFIG_XEN_PVHVM)		+= enlighten_hvm.o mmu_hvm.o suspend_hvm.o
+ obj-$(CONFIG_XEN_PV)			+= setup.o apic.o pmu.o suspend_pv.o \
+@@ -32,5 +32,4 @@ obj-$(CONFIG_PARAVIRT_SPINLOCKS)+= spinl
+ obj-$(CONFIG_XEN_DEBUG_FS)	+= debugfs.o
+ obj-$(CONFIG_XEN_DOM0)		+= vga.o
+ obj-$(CONFIG_SWIOTLB_XEN)	+= pci-swiotlb-xen.o
+-obj-$(CONFIG_XEN_EFI)		+= efi.o
+ obj-$(CONFIG_XEN_PVH)	 	+= xen-pvh.o
+--- a/arch/x86/xen/efi.c
++++ b/arch/x86/xen/efi.c
+@@ -127,9 +127,8 @@ void __init xen_efi_init(void)
+ 	strncpy((char *)&boot_params.efi_info.efi_loader_signature, "Xen",
+ 			sizeof(boot_params.efi_info.efi_loader_signature));
+ 	boot_params.efi_info.efi_systab = (__u32)__pa(efi_systab_xen);
+-	boot_params.efi_info.efi_systab_hi = (__u32)(__pa(efi_systab_xen) >> 32);
++	boot_params.efi_info.efi_systab_hi = 0;
```

Comment:
This looks bogus.  efi_systab_xen is a linux data structure.  In a 32bit build, it can't live above 4GB, so it should be unnecessary.

Reply:
On a 32 bit arch, shifts of 32 bit size are undefined, sometimes they return the original variable unshifted. (ie. 5 >> 32 == 5).  I set it to 0 just to be safe.

Reply:
Ah, I naively assumed __pa was returning a 64bit value.  So it's a bug in linux.

Reply:
Yea I suppose they never considered the case of running the xen_efi_init codepath on a 32-bit dom0.